packages/teleterm/src/mainProcess/protocolHandler.ts

```diff
@@ -0,0 +1,88 @@
+import type {
+  protocol as ElectronProtocol,
+  ProtocolRequest,
+  ProtocolResponse,
+} from 'electron';
+import * as path from 'path';
+import fs from 'fs';
+
+type CallbackType = (response: string | ProtocolResponse) => void;
+
+function _disabledHandler(
+  _request: ProtocolRequest,
+  callback: CallbackType
+): void {
+  callback({ error: -3 });
+}
+
+export function installWebHandler({
+  protocol,
+}: {
+  protocol: typeof ElectronProtocol;
+}): void {
+  protocol.interceptFileProtocol('about', _disabledHandler);
+  protocol.interceptFileProtocol('content', _disabledHandler);
+  protocol.interceptFileProtocol('chrome', _disabledHandler);
+  protocol.interceptFileProtocol('cid', _disabledHandler);
+  protocol.interceptFileProtocol('data', _disabledHandler);
+  protocol.interceptFileProtocol('filesystem', _disabledHandler);
+  protocol.interceptFileProtocol('ftp', _disabledHandler);
+  protocol.interceptFileProtocol('gopher', _disabledHandler);
+  protocol.interceptFileProtocol('javascript', _disabledHandler);
+  protocol.interceptFileProtocol('mailto', _disabledHandler);
+}
+
+function eliminateAllAfterCharacter(string, character) {
+  const index = string.indexOf(character);
+  if (index < 0) {
+    return string;
+  }
+
+  return string.slice(0, index);
+}
+
+function urlToPath(targetUrl, { isWin }) {
+  const decoded = decodeURIComponent(targetUrl);
+  const withoutScheme = decoded.slice(isWin ? 8 : 7);
+
+  // remove any extra characters in requsted file path
+  const withoutQuerystring = eliminateAllAfterCharacter(withoutScheme, '?');
+  const withoutHash = eliminateAllAfterCharacter(withoutQuerystring, '#');
+
+  return withoutHash;
+}
+
+// intercept, clean, and validate the requested file path.
+export function interceptFileProtocol({
+  protocol,
+  isWin,
+  userDataPath,
+  installPath,
+  logger,
+}) {
+  protocol.interceptFileProtocol('file', (request, callback) => {
+    const target = path.normalize(urlToPath(request.url, { isWin }));
+    const realPath = fs.existsSync(target) ? fs.realpathSync(target) : target;
+    const properCasing = isWin ? realPath.toLowerCase() : realPath;
+    if (
+      !properCasing.startsWith(
+        isWin ? userDataPath.toLowerCase() : userDataPath
+      ) &&
```

Comment:
What's the reason behind having to call `toLowerCase` on Windows?

What do you think about moving those transformations out of the conditional, so we'd have something like:

```typescript
const userDataPath = isWin ? userDataPath.toLowerCase() : userDataPath
```

Reply:
unix is sensitive and windows isn't when it comes to file names. So this was just to make sure the check was always happening on the same casing but... I suppose it isn't needed. a few things in this PR were 'super duper safety checks'.

I'm down to keep it but to move it out of the conditional 👍

Reply:
Makes sense, I'd also write that down in a comment.